docs/ref-manual/cvl/methods.md

```diff
@@ -46,44 +55,251 @@ method_summary   ::= "ALWAYS" "(" value ")"
+statements <using>` can also be described by prefixing the method name with
+the contract variable name.  For example, if contract `C` is introduced by the
+statement `using C as c`, then the method `f(uint)` of contract `c` can be
+referred to as `c.f(uint)`.
+
+It is possible for a method signature to appear in the `methods` block but not
+in the contract being verified.  In this case, the prover will skip any rules
+that mention the missing method, rather than reporting an error.  This behavior
+allows reusing specifications on contracts that only support part of an
+interface: only the supported methods will be verified.
+
+Following the method signature is an optional `returns` clause.  If a method
+declaration contains a `returns` clause, the declared return type must match
+the contract method's return type.  If the `returns` clause is omitted, the
+return type is taken from the contract method's return type.
+
+Following the `returns` clause is an optional `envfree` tag.  Marking a method
+with `envfree` has two effects.  First, {ref}`calls <call-expr>` to the method
+from CVL do not need to explicitly pass an {term}`environment` value as the
+first argument.  Second, the prover will verify that the method implementation
+in the contract being verified does not depend on any of the environment
+variables.  The results of this check are displayed on the verification report
+as separate rules called `envfreeFuncsStaticCheck` and
+`envfreeFuncsAreNonpayable`[^envfree_nonpayable].
+
+[^envfree_nonpayable]: The effect of payable functions on the contract's
+  balance depends on the message value, so payable functions also require an
+  `env`.
+
+Finally, the method entry may contain an optional summarization (indicated by
+`=>` followed by the summary type and an optional application policy).  A
+summarized declaration indicates that the prover should replace some calls to
+the summarized function by an approximation.  This is an important technique
+for working around prover timeouts and also for working with external contracts
+whose implementation is not fixed at verification time.
+
+The summary type determines what type of approximation is used to replace the
+function calls.  The available types are described in the following sections:
+
+ * {ref}`const-summary`
+ * {ref}`havoc-summary`
+ * {ref}`dispatcher`
+ * {ref}`auto-summary`
+ * {ref}`function-summary`
+
+The application policy determines which function calls are replaced by
+approximations.  See {ref}`summaries` for details.
+
+(summaries)=
+Which function calls are summarized
+-----------------------------------
+
+Method summaries apply to all calls with the matching ABI signature, including
+internal methods and external methods on all contracts.  There is currently no
+way to apply different summaries to different contracts or to summarize some
+calls and not others to methods with the same ABI signature.  For this reason,
```

Comment:
this is not, strictly speaking, true. It is possible that for some function, e.g., `owner()`, the callee for some call-sites is known and for others it is not. If you don't specify an forced summary, then the summary behavior will only apply for those callsites where the callee is unresolved.

Reply:
I think this behavior is described below, but I should be more careful here to distinguish between which methods match and which methods are approximated.

Reply:
> this is not, strictly speaking, true. It is possible that for some function, e.g., owner(), the callee for some call-sites is known and for others it is not. If you don't specify an forced summary, then the summary behavior will only apply for those callsites where the callee is unresolved.

I think this is not quite what we're talking about here. Perhaps there should be a more clear distinction in language between how summaries are "specified" and how the tool "applies" them.

Reply:
I think the latest draft clarifies this